app/Models/BeatmapPack.php

```diff
@@ -22,9 +22,67 @@
 
 class BeatmapPack extends Model
 {
+    const DEFAULT_TYPE = 'standard';
+    private static $tagMappings = [
+        'standard' => 'S',
+        'theme' => 'T',
+        'artist' => 'A',
+        'chart' => 'R',
+    ];
+
     protected $table = 'osu_beatmappacks';
     protected $primaryKey = 'pack_id';
 
     protected $dates = ['date'];
     public $timestamps = false;
+
+    public function items()
+    {
+        return $this->hasMany(BeatmapPackItem::class, 'pack_id');
+    }
+
+    public function beatmapsets()
+    {
+        $thisTable = $this->getTable();
+        $setsTable = (new Beatmapset)->getTable();
+        $itemsTable = (new BeatmapPackItem)->getTable();
+
+        return Beatmapset::query()
+            ->join($itemsTable, "{$itemsTable}.beatmapset_id", '=', "{$setsTable}.beatmapset_id")
+            ->join($thisTable, "{$itemsTable}.pack_id", '=', "{$thisTable}.pack_id")
```

Comment:
this join not needed (and change the table name down there)?

(or just `whereIn('beatmapset_id', $this->items()->pluck('pack_id'))` with one extra query)